### PR TITLE
[SPARK-7132][ML] Add fit with validation set to spark.ml GBT

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

Add fit with validation set to spark.ml GBT

## How was this patch tested?

Will add later.

### CODE REVIEW

**Comment by WeichenXu123** [comment]
Jenkins, test this please.

File: mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala [hunk 1]
```diff
@@ -95,7 +95,9 @@ private[shared] object SharedParamsCodeGen {
       ParamDesc[String]("distanceMeasure", "The distance measure. Supported options: 'euclidean'" +
         " and 'cosine'", Some("org.apache.spark.mllib.clustering.DistanceMeasure.EUCLIDEAN"),
         isValid = "(value: String) => " +
-        "org.apache.spark.mllib.clustering.DistanceMeasure.validateDistanceMeasure(value)")
+        "org.apache.spark.mllib.clustering.DistanceMeasure.validateDistanceMeasure(value)"),
+      ParamDesc[String]("validationIndicatorCol", "the indicator column name for indicating " +
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/param/shared/SharedParamsCodeGen.scala]: How about rephrasing the description: "name of the column that indicates whether each row is for training or for validation.  False indicates training; true indicates validation."

**Review by jkbradley — COMMENTED** [review]
Could you please add unit tests?  I'll continue with a detailed review afterwards.

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala [hunk 2]
```diff
@@ -365,6 +366,50 @@ class GBTClassifierSuite extends MLTest with DefaultReadWriteTest {
...
+        .setMaxDepth(2)
+        .setLossType(lossType)
+        .setMaxIter(numIter)
+      val modelWithoutValidation = gbt.fit(trainDF)
+
+      gbt.setValidationIndicatorCol(validationIndicatorCol)
+      val modelWithValidation = gbt.fit(trainDF.union(validationDF))
+
+      // early stop
+      assert(modelWithValidation.numTrees < numIter)
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: Let's also assert `modelWithoutValidation.numTrees == numIter`.  That's true now, but I could imagine it changing later on if we add a convergence tolerance to the algorithm.

File: mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala [hunk 3]
```diff
@@ -365,6 +366,50 @@ class GBTClassifierSuite extends MLTest with DefaultReadWriteTest {
...
+
+      val (errorWithoutValidation, errorWithValidation) = {
+        val remappedRdd = validationData.map(x => new LabeledPoint(2 * x.label - 1, x.features))
+        (GradientBoostedTrees.computeError(remappedRdd, modelWithoutValidation.trees,
+          modelWithoutValidation.treeWeights, modelWithoutValidation.getOldLossType),
+          GradientBoostedTrees.computeError(remappedRdd, modelWithValidation.trees,
+            modelWithValidation.treeWeights, modelWithValidation.getOldLossType))
+      }
+      assert(errorWithValidation <= errorWithoutValidation)
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/classification/GBTClassifierSuite.scala]: It'd be nice to have this be strictly true. Is it not?

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala [hunk 4]
```diff
@@ -497,6 +498,9 @@ private[ml] trait GBTParams extends TreeEnsembleParams with HasMaxIter with HasS
   @deprecated("This method is deprecated and will be removed in 3.0.0.", "2.1.0")
   def setStepSize(value: Double): this.type = set(stepSize, value)
 
+  /** @group setParam */
+  def setValidationIndicatorCol(value: String): this.type = set(validationIndicatorCol, value)
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: Since we need this setter to be in the concrete classes for Java compatibility, don't bother putting it here.

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala [hunk 5]
```diff
@@ -146,20 +146,40 @@ class GBTClassifier @Since("1.4.0") (
...
+  @Since("2.4.0")
+  override def setValidationIndicatorCol(value: String): this.type = {
+    set(validationIndicatorCol, value)
+  }
+
   override protected def train(dataset: Dataset[_]): GBTClassificationModel = {
     val categoricalFeatures: Map[Int, Int] =
       MetadataUtils.getCategoricalFeatures(dataset.schema($(featuresCol)))
+
+    val withValidation = isDefined(validationIndicatorCol)
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: Let's do: `isDefined(validationIndicatorCol) and $(validationIndicatorCol).nonEmpty` (so we count "" as not defined).

File: mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala [hunk 6]
```diff
@@ -146,20 +146,40 @@ class GBTClassifier @Since("1.4.0") (
...
+
     // We copy and modify this from Classifier.extractLabeledPoints since GBT only supports
     // 2 classes now.  This lets us provide a more precise error message.
-    val oldDataset: RDD[LabeledPoint] =
+    val convert2LabelPoint = (dataset: Dataset[_]) => {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala]: nit: "LabelPoint" -> "LabeledPoint"

**Review by jkbradley — COMMENTED** [review]
Thanks for the unit tests!

Let's keep the validationTol Param from the spark.mllib API.  That's a pretty important parameter for users.

**Comment by WeichenXu123** [comment]
Jenkins, test this please.

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala [hunk 7]
```diff
@@ -460,18 +461,29 @@ private[ml] trait RandomForestRegressorParams
...
-  /* TODO: Add this doc when we add this param.  SPARK-7132
-   * Threshold for stopping early when runWithValidation is used.
+  /**
+   * Threshold for stopping early when fit with validation is used.
    * If the error rate on the validation input changes by less than the validationTol,
-   * then learning will stop early (before [[numIterations]]).
-   * This parameter is ignored when run is used.
+   * then learning will stop early (before [[maxIter]]).
+   * This parameter is ignored when fit without validation is used.
    * (default = 1e-5)
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: I forget why we chose 1e-5 (which is different from spark.mllib).  What do you think about using 0.01 to match the sklearn docs here? http://scikit-learn.org/dev/auto_examples/ensemble/plot_gradient_boosting_early_stopping.html  (I also checked xgboost, but they use a different approach based on x number of steps without improvement.  We may want to add that at some point since it sounds more robust.)

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala [hunk 8]
```diff
@@ -460,18 +461,29 @@ private[ml] trait RandomForestRegressorParams
...
    * @group param
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: Let's add a `@see` validationIndicatorCol

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala [hunk 9]
```diff
@@ -460,18 +461,29 @@ private[ml] trait RandomForestRegressorParams
...
  */
-private[ml] trait GBTParams extends TreeEnsembleParams with HasMaxIter with HasStepSize {
+private[ml] trait GBTParams extends TreeEnsembleParams with HasMaxIter with HasStepSize
+  with HasValidationIndicatorCol {
 
-  /* TODO: Add this doc when we add this param.  SPARK-7132
-   * Threshold for stopping early when runWithValidation is used.
+  /**
+   * Threshold for stopping early when fit with validation is used.
    * If the error rate on the validation input changes by less than the validationTol,
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: Let's add the more precise description from the old BoostingStrategy for this Param.

**Review by jkbradley — COMMENTED** [review]
Thanks for the updates!

**Comment by WeichenXu123** [comment]
Jenkins test this please.

File: mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala [hunk 10]
```diff
@@ -460,18 +461,37 @@ private[ml] trait RandomForestRegressorParams
...
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala]: This doc is not quite accurate.  Can you please update it to:
```
Threshold for stopping early when fit with validation is used.
(This parameter is ignored when fit without validation is used.)
The decision to stop early is decided based on this logic:
If the current loss on the validation set is greater than 0.01, the diff
of validation error is compared to relative tolerance which is
validationTol * (current loss on the validation set).
If the current loss on the validation set is less than or equal to 0.01,
the diff of validation error is compared to absolute tolerance which is
validationTol * 0.01.
```

**Comment by jkbradley** [comment]
LGTM
Merging with master
Thanks!

**Comment by jkbradley** [comment]
I made a JIRA for the Python part of this: https://issues.apache.org/jira/browse/SPARK-24333
